text/0064-asset-garbage-collection.md

```diff
@@ -0,0 +1,171 @@
+* **Tracking Issue**: #64
+* **API Bar Raiser**: @nlynch
+
+The asset garbage collection CLI will identify and/or delete unused CDK assets,
+resulting in smaller bucket/repo size and less cost for customers.
+
+## Working Backwards
+
+**CHANGELOG**:
+
+- feat(cli): `cdk gc` - garbage collection of unused assets
+
+**Help**:
+
+```shell
+➜ cdk gc --help
+cdk gc [ENVIRONMENT...]
+
+Finds and deletes all unused S3 and ECR assets in the ENVIRONMENT
+
+Options:
+  -l, --list              lists, rather than deletes, all unused assets
+  -t, --type=[s3|ecr]     filters for type of asset
+
+Examples:
+  cdk gc
+  cdk gc aws://ACCOUNT/REGION
+  cdk gc --list --type=s3
+```
+
+**README:**
+
+The `cdk gc` command reduces the size of your S3 asset bucket and your ecr asset
+repository by identifying and deleting unused assets. The command will enter the
+specified environment or all environments in the CDK app and determine assets that are no
+longer referenced and can be safely deleted. You can also list unused assets using the
+`--list` option or filter for a specific type of asset using `--type`.
+
+**Usage:**
+
+This command will garbage collect all unused assets in all environments that belong to the current CDK app (if `cdk.json` exists):
+
+```shell
+cdk gc
+```
+
+This command will list all unused assets in all environments that belong to the current CDK app:
+
+```shell
+cdk gc --list
+```
+
+This command will garbage collect unused assets in the specified environment:
+
+```shell
+cdk gc aws://ACCOUNT/REGION
+```
+
+This command will list unused S3 assets in the specified environment. The options for `--type` are `s3` or `ecr`:
+
+```shell
+cdk gc aws://ACOUNT/REGION --type=s3
+```
+
+---
+
+#
+
+Ticking the box below indicates that the public API of this RFC has been
+signed-off by the API bar raiser (the `api-approved` label was applied to the
+RFC pull request):
+
+```
+[ ] Signed-off by API Bar Raiser @xxxxx
+```
+
+## Public FAQ
+
+> This section should include answers to questions readers will likely ask about
+> this release. Similar to the "working backwards", this section should be
+> written in a language as if the feature is now released.
+>
+> The template includes a some common questions, feel free to add any questions
+> that might be relevant to this feature or omit questions that you feel are not
+> applicable.
+
+### What are we launching today?
+
+The `cdk gc` command features, with support for garbage collection of unused S3 and ECR
+assets.
+
+### Why should I use this feature?
+
+Currently unused assets are left in the S3 bucket or ECR repository and contribute
+additional cost for customers. This feature provides a swift way to identify and delete
+unutilized assets.
+
+### How does the command identify unused assets?
+
+`cdk gc` will look at all the deployed, healthy stacks in the environment and trace the
+assets that are being referenced by these stacks. All assets that are not reached via
+tracing can be safely deleted.
```

Comment:
Something to think about which I think is not optional: how will we leave a safety margin for rollbacks?

As an example:

* A pipeline deploys revision R1 with some version of some Lambda code (`r1.zip`)
* A pipeline deploys revision R2, changing the Lambda code (`r2.zip`)
* Garbage collection gets run (deleting `r1.zip`)
* For whatever reason, the app team wants to rollback to R1, but `r1.zip` is now gone

I guess the answer to what we need to do in GC depends on how rollback is going to work. If it's going to be a full `cdk deploy` cycle we could rely on the CDK CLI to re-upload the assets... but how do CD systems work?

* In a CD system like Amazon uses internally the CFN deployment would be executed, but the asset deployment wouldn't (and so `r1.zip` would be gone without recovery)
* CodePipeline doesn't currently support rollbacks, but it could in the future (and chances are it would work like Amazon's does)

All in all, you need to think about this case and think about what is reasonable to expect from a CI/CD system and what is reasonable for us to implement, and write a couple words about that.

For example, a reasonable safety precaution would be we only delete assets 30 days (or pick a number) after they stop being used. That would allow for safe interoperability even with pessimistic CI/CD scenarios.

Reply:
This is a good idea and something I can integrate into the RFC. I'll start with 30 days, since I'm not all that familiar with how people use cloudformation rollbacks I don't have an idea of what (if any) is a more suitable time period.

Reply:
[For this](https://github.com/aws/aws-cdk-rfcs/pull/379#discussion_r704698965)
> This is a good idea and something I can integrate into the RFC. I'll start with 30 days, since I'm not all that familiar with how people use cloudformation rollbacks I don't have an idea of what (if any) is a more suitable time period.

I don't think this can entirely be solved with timestamps. As I understand there are two kinds of race conditions:

1. You fetch the templates when there were some in-progress stacks. If you deleted the assets cross-referencing such templates you might delete assets that would be needed if one or more of these in-progress stacks rolled-back (because the `CFn` template got would be the new one which was cause of the update). This cannot be remedied with comparing asset timestamps as it's orthogonal to that - it's about the 2 versions rather than two timestamps.
2. You could also race with asset-uploads. If an asset got uploaded but the changeset wasn't executed yet and your workflow ran and referenced the about to be updated template, then you would delete the asset that just got uploaded and the stack update would fail. This is better than a rollback failure of (1) as it'll be an update failure - but still a failure.

I think (2) can be remedied by using a timestamp - leave alone all assets not older than X days where X is the max likely time difference in someone's pipeline between the asset-upload stage and changeset execution stage.

For (1), assuming that is correct, there's some atomicity needed. I guess if we could atomically ask for stack-status+template in the same call, we could decide to run the workflow only if none of the stacks were in `in-progress` state (as that can rollback to a template which you don't know about). Then combine it with the timestamp check of (2) so that it doesn't race with newer deployments. This will still race with `create-in-progress` as such a stack might not even have existed during the scan and you will end up deleting the assets of about to be created stack. Again age-check of (2) could remedy this. If there are multiple pipelines (thus multiple stacks) deploying to the same account+region, you will need to configure `X` to be the worst case max among all the pipelines reaching that stage (obviously - but just stating).

---

Probably not going to happen, but if we had per stack asset bucket-prefix (or entire bucket) and ecr, then we could simplify all this in case of pipeline deployments by having a post-deployment action that cleans all these assets blindly by referring to the just succeeded stack. We don't have to bother about other stacks in the account + it'll only happen after a successful update of a stack (once the stack is "stable") - so no need to worry about in-progress states. This combined with the "age" workaround might have been the best? The drawback is if you have several stacks/templates that actually reference the same asset, that would lead to asset duplication - but maybe that's OK?